Compilation errors on the 4.14.141 kernel version

Hi @menglongdong 

Sorry to bother you again :(

When attempting to compile this tool on the 4.14.141 kernel version, I encountered the following errors:

```
$ make -C $(pwd) COMPAT=1 all

make: Entering directory '/tmp1/nettrace'
make -C src all
make[1]: Entering directory '/tmp1/nettrace/src'
python3 gen_trace.py > trace_group.c
python3 gen_trace.py probe > progs/kprobe_trace.h
ln -s vmlinux_header.h kheaders.h
clang -O2 -c -S -Wall -fno-asynchronous-unwind-tables		\
-Wno-incompatible-pointer-types-discards-qualifiers		\
progs/kprobe.c -emit-llvm -Wno-unknown-attributes -I./ -I/tmp1/nettrace/shared/bpf/ -g -DBPF_NO_GLOBAL_DATA -DNO_BTF -DINLINE_MODE -D__F_STACK_TRACE -D__KERN_VER=4.14.141 -D__KERN_MAJOR=4 -Wno-unused-function -Wno-compare-distinct-pointer-types -Wuninitialized -D__TARGET_ARCH_x86 -DBPF_NO_PRESERVE_ACCESS_INDEX -nostdinc -isystem /usr/lib/gcc/x86_64-linux-gnu/12/include -I/lib/modules/4.14.141/build//arch/x86/include -I/lib/modules/4.14.141/build//arch/x86/include/generated -I/lib/modules/4.14.141/build//include -I/lib/modules/4.14.141/build//arch/x86/include/uapi -I/lib/modules/4.14.141/build//arch/x86/include/generated/uapi -I/lib/modules/4.14.141/build//include/uapi -I/lib/modules/4.14.141/build//include/generated/uapi -include /lib/modules/4.14.141/build//include/linux/kconfig.h -I/usr/include/ -D__KERNEL__ -Wno-unused-value -Wno-pointer-sign -Wno-compare-distinct-pointer-types -Wno-gnu-variable-sized-type-not-at-end -Wno-address-of-packed-member -Wno-tautological-compare -Wno-unknown-warning-option -Wno-frame-address -D__F_SK_PRPTOCOL_LEGACY -Xclang	\
-disable-llvm-passes -o - | 					\
opt -O2 -mtriple=bpf-pc-linux | 				\
llvm-dis |							\
llc -march=bpf -filetype=obj -o progs/kprobe.o
In file included from progs/kprobe.c:3:
In file included from /usr/include/bpf/bpf_helpers.h:11:
/usr/include/bpf/bpf_helper_defs.h:327:15: error: redefinition of 'bpf_tail_call' as different kind of symbol
static long (*bpf_tail_call)(void *ctx, void *prog_array_map, __u32 index) = (void *) 12;
              ^
/lib/modules/4.14.141/build//include/linux/bpf.h:236:5: note: previous definition is here
u64 bpf_tail_call(u64 ctx, u64 r2, u64 index, u64 r4, u64 r5);
    ^
In file included from progs/kprobe.c:3:
In file included from /usr/include/bpf/bpf_helpers.h:11:
/usr/include/bpf/bpf_helper_defs.h:763:15: error: redefinition of 'bpf_get_stackid' as different kind of symbol
static long (*bpf_get_stackid)(void *ctx, void *map, __u64 flags) = (void *) 27;
              ^
/lib/modules/4.14.141/build//include/linux/bpf.h:237:5: note: previous definition is here
u64 bpf_get_stackid(u64 r1, u64 r2, u64 r3, u64 r4, u64 r5);
    ^
In file included from progs/kprobe.c:94:
progs/core.c:57:60: error: too few arguments to function call, expected 5, have 3
        info->e->stack_id = bpf_get_stackid(info->ctx, &m_stack, 0);
                            ~~~~~~~~~~~~~~~                       ^
/lib/modules/4.14.141/build//include/linux/bpf.h:237:5: note: 'bpf_get_stackid' declared here
u64 bpf_get_stackid(u64 r1, u64 r2, u64 r3, u64 r4, u64 r5);
    ^
3 errors generated.
make[1]: *** [../common.mk:112: progs/kprobe.o] Error 1
make[1]: Leaving directory '/tmp1/nettrace/src'
make: *** [Makefile:24: all] Error 2
make: Leaving directory '/tmp1/nettrace'
```

It's always tough when we want eBPF tools to be compatible with as many kernel versions as possible ;)

--- a/.gitignore
+++ b/.gitignore
@@ -11,3 +11,4 @@ GRTAGS
 kheaders.h
 build*.sh
 release.sh
+compile_commands.json

--- a/component/arg_parse.c
+++ b/component/arg_parse.c
@@ -5,12 +5,14 @@
 #include <string.h>
 #include <stdlib.h>
 #include <inttypes.h>
+#include <errno.h>
 #define _LINUX_IN_H
 #include <netinet/in.h>
 #include <arpa/inet.h>
 
 #include "arg_parse.h"
 #include "net_utils.h"
+#include "sys_utils.h"
 
 #define KEY_START	1000
 

--- a/component/arg_parse.h
+++ b/component/arg_parse.h
@@ -4,8 +4,6 @@
 #define _H_ARG_PARSE
 #include <stdbool.h>
 
-#include "sys_utils.h"
-
 enum option_type {
 	OPTION_STRING,
 	OPTION_BOOL,

--- a/component/net_utils.c
+++ b/component/net_utils.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: MulanPSL-2.0
 
 #include "net_utils.h"
-#include "arg_parse.h"
+#include "sys_utils.h"
 
 #define _LINUX_IN_H
 #include <netinet/in.h>

--- a/component/net_utils.h
+++ b/component/net_utils.h
@@ -5,14 +5,9 @@
 
 #include <unistd.h>
 #include <getopt.h>
-#include <signal.h>
 #include <stdbool.h>
-#include <stdio.h>
-#include <string.h>
 #include <sys/types.h>
 #include <asm-generic/int-ll64.h>
-#include <stdlib.h>
-#include <errno.h>
 #include <bpf/libbpf.h>
 #include <linux/if_ether.h>
 

--- a/src/analysis.c
+++ b/src/analysis.c
@@ -8,6 +8,7 @@
 #undef __USE_MISC
 #include <net/if.h>
 #include <pthread.h>
+#include <errno.h>
 
 #include <pkt_utils.h>
 #include <stdlib.h>

--- a/src/dropreason.h
+++ b/src/dropreason.h
@@ -1,6 +1,8 @@
 #ifndef _H_SKB_DROP_REASON
 #define _H_SKB_DROP_REASON
 
+#include <stdbool.h>
+
 char *get_drop_reason(int index);
 bool drop_reason_support();
 

--- a/src/nettrace.c
+++ b/src/nettrace.c
@@ -1,6 +1,8 @@
 // SPDX-License-Identifier: MulanPSL-2.0
 
+#include <errno.h>
 #include <arpa/inet.h>
+#include <signal.h>
 
 #include <arg_parse.h>
 

--- a/src/trace.c
+++ b/src/trace.c
@@ -2,10 +2,10 @@
 
 #include <stdio.h>
 #include <list.h>
+#include <errno.h>
 
 #include <parse_sym.h>
 
-#include "nettrace.h"
 #include "trace.h"
 #include "analysis.h"
 #include "dropreason.h"

--- a/src/trace_probe.c
+++ b/src/trace_probe.c
@@ -4,8 +4,6 @@
 #include "trace.h"
 #include "progs/kprobe.skel.h"
 #include "analysis.h"
-#include "nettrace.h"
-#include "analysis.h"
 
 #define MAX_CPU_COUNT 1024
 

--- a/src/trace_tracing.c
+++ b/src/trace_tracing.c
@@ -3,8 +3,6 @@
 
 #include "trace.h"
 #include "analysis.h"
-#include "nettrace.h"
-#include "analysis.h"
 
 /* check whether trampoline is supported by current arch */
 static bool tracing_arch_supported()

--- a/src/vmlinux_header.h
+++ b/src/vmlinux_header.h
@@ -16,7 +16,12 @@
 #include <net/sch_generic.h>
 #include <net/sock.h>
 #include <net/inet_connection_sock.h>
+
+#define bpf_tail_call dummy____bpf_tail_call
+#define bpf_get_stackid dummy____bpf_get_stackid
 #include <net/tcp.h>
+#undef bpf_tail_call
+#undef bpf_get_stackid
 
 #if defined(CONFIG_NF_TABLES) || defined(CONFIG_NF_TABLES_MODULE)
 #include <net/netfilter/nf_tables.h>

